### PR TITLE
BaSP-MR-31: SUBSCRIPTIONS - get, getById and create Test

### PR DESCRIPTION
**Pull request by jramire5**
Here you have some tests to check if the Subscriptions feature is working well, I invite you to take a look, and find as many corrections as possible, any comment is welcomed.

### CODE REVIEW

File: src/tests/subscriptions.test.js [hunk 1]
```diff
@@ -1,5 +1,121 @@
...
+describe('GET /api/subscriptions', () => {
+  test('should return status 200', async () => {
+    const response = await request(app).get(validRoute).send();
+    expect(response.status).toBe(200);
+    expect(response.body.data).toBeDefined();
+    expect(response.body.error).toBe(false);
+  });
+  test('should return status 404 and no subscriptions message', async () => {
+    await Subscription.collection.deleteMany();
+    const response = await request(app).get('/api/subscriptions').send();
```
Comment (AleMac17) [inline src/tests/subscriptions.test.js:37]: You can use the const valid route that you defined before inside the get
`get(validRoute)`

File: src/tests/subscriptions.test.js [hunk 2]
```diff
@@ -1,5 +1,121 @@
...
+    expect(response.status).toBe(404);
+    expect(response.body.data).toBeDefined();
+    expect(response.body.error).toBeTruthy();
+    expect(response.body.message).toBe('There are no subscriptions!');
+  });
+  test('should return status 500', async () => {
+    jest.spyOn(Subscription, 'populate').mockRejectedValue(new Error('oh no Something went wrong'));
+    const response = await request(app).get('/api/subscriptions/').send();
```
Comment (AleMac17) [inline src/tests/subscriptions.test.js:45]: You can use the const valid route that you defined before inside the get
`get(validRoute)`

File: src/tests/subscriptions.test.js [hunk 3]
```diff
@@ -1,5 +1,121 @@
...
+  test('should return status 500', async () => {
+    jest.spyOn(Subscription, 'populate').mockRejectedValue(new Error('Something went wrong'));
+    const response = await request(app).get(`/api/subscriptions/${mockSubscriptionId}`).send();
+    expect(response.status).toBe(500);
+  });
+});
+
+describe('POST /api/subscriptions/', () => {
+  test('should return status 201', async () => {
+    const response = await request(app).post('/api/subscriptions/').send(mockSubscription);
```
Comment (AleMac17) [inline src/tests/subscriptions.test.js:81]: You can use the const valid route that you defined before inside the get
`get(validRoute)`

File: src/tests/subscriptions.test.js [hunk 4]
```diff
@@ -1,5 +1,121 @@
...
+    expect(response.status).toBe(201);
+    expect(response.body.data).toBeDefined();
+    expect(response.body.error).toBeFalsy();
+  });
+  test('should return status 400 and missing date message', async () => {
+    const response = await request(app).post('/api/subscriptions/').send({
```
Comment (AleMac17) [inline src/tests/subscriptions.test.js:87]: You can use the const valid route that you defined before inside the get
`get(validRoute)`

File: src/tests/subscriptions.test.js [hunk 5]
```diff
@@ -1,5 +1,121 @@
...
+      classId: mockSubscription.classId,
+      date: mockSubscription.date,
+    });
+    expect(response.status).toBe(400);
+    expect(response.body.message).toBe('An error has occurred: "members" is required');
+    expect(response.body.data).toBeUndefined();
+    expect(response.body.error).toBeTruthy();
+  });
+  test('should return status 400 and missing class Id message', async () => {
+    const response = await request(app).post('/api/subscriptions/').send({
```
Comment (AleMac17) [inline src/tests/subscriptions.test.js:107]: You can use the const valid route that you defined before inside the get
`get(validRoute)`

File: src/tests/subscriptions.test.js [hunk 6]
```diff
@@ -1,5 +1,121 @@
...
+      date: mockSubscription.date,
+    });
+    expect(response.status).toBe(400);
+    expect(response.body.data).toBeUndefined();
+    expect(response.body.message).toBe('An error has occurred: "classId" is required');
+    expect(response.body.error).toBeTruthy();
+  });
+  test('should return status 500', async () => {
+    jest.spyOn(Subscription, 'create').mockRejectedValue(new Error('Oh no! Something went wrong!'));
+    const response = await request(app).post('/api/subscriptions/').send(mockSubscription);
```
Comment (AleMac17) [inline src/tests/subscriptions.test.js:118]: You can use the const valid route that you defined before inside the get
`get(validRoute)`

File: src/tests/subscriptions.test.js [hunk 7]
```diff
@@ -1,5 +1,121 @@
...
+  await Subscription.collection.deleteMany();
+  jest.restoreAllMocks();
+});
+
+describe('GET /api/subscriptions', () => {
+  test('should return status 200', async () => {
+    const response = await request(app).get(validRoute).send();
+    expect(response.status).toBe(200);
+    expect(response.body.data).toBeDefined();
+    expect(response.body.error).toBe(false);
```
Comment (JuanManuel-Blois) [inline src/tests/subscriptions.test.js:33]: we should add another expect for the message

File: src/tests/subscriptions.test.js [hunk 8]
```diff
@@ -1,5 +1,121 @@
...
+    expect(response.status).toBe(500);
+  });
+});
+
+describe('POST /api/subscriptions/', () => {
+  test('should return status 201', async () => {
+    const response = await request(app).post('/api/subscriptions/').send(mockSubscription);
+    expect(response.status).toBe(201);
+    expect(response.body.data).toBeDefined();
+    expect(response.body.error).toBeFalsy();
```
Comment (JuanManuel-Blois) [inline src/tests/subscriptions.test.js:84]: we should add another expect to check the message

File: src/tests/subscriptions.test.js [hunk 9]
```diff
@@ -1,5 +1,124 @@
...
+    expect(response.status).toBe(500);
+  });
+});
+
+describe('POST /api/subscriptions/', () => {
+  test('should return status 201', async () => {
+    const response = await request(app).post(validRoute).send(mockSubscription);
+    expect(response.status).toBe(201);
+    expect(response.body.message).toBe('Subscription created successfully!');
+
```
Comment (JuanManuel-Blois) [inline src/tests/subscriptions.test.js:85]: there's an extra empty line here

**Review by JuanManuel-Blois — APPROVED** [review]

File: src/tests/subscriptions.test.js [hunk 10]
```diff
@@ -1,5 +1,123 @@
...
+    const response = await request(app).get(`/api/subscriptions/${notFoundId}`).send();
+    expect(response.status).toBe(404);
+    expect(response.body.data).toBeDefined();
+    expect(response.body.message).toBe(`Subscription with id: ${notFoundId} not found!`);
+    expect(response.body.error).toBeTruthy();
+  });
+  test('should return status 500', async () => {
+    jest.spyOn(Subscription, 'populate').mockRejectedValue(new Error('Something went wrong'));
+    const response = await request(app).get(`/api/subscriptions/${mockSubscriptionId}`).send();
+    expect(response.status).toBe(500);
```
Comment (JuanManuel-Blois) [inline src/tests/subscriptions.test.js:76]: we should check for the data, message and error in the body

File: src/tests/subscriptions.test.js [hunk 11]
```diff
@@ -1,5 +1,123 @@
...
+    });
+    expect(response.status).toBe(400);
+    expect(response.body.data).toBeUndefined();
+    expect(response.body.message).toBe('An error has occurred: "classId" is required');
+    expect(response.body.error).toBeTruthy();
+  });
+  test('should return status 500', async () => {
+    jest.spyOn(Subscription, 'create').mockRejectedValue(new Error('Oh no! Something went wrong!'));
+    const response = await request(app).post(validRoute).send(mockSubscription);
+    expect(response.status).toBe(500);
```
Comment (JuanManuel-Blois) [inline src/tests/subscriptions.test.js:121]: we should check for the data, message and error in the body

File: src/tests/subscriptions.test.js [hunk 12]
```diff
@@ -1,5 +1,123 @@
...
+    expect(response.status).toBe(200);
+    expect(response.body.data).toBeDefined();
+    expect(response.body.message).toBe('Subscriptions found successfully!');
+    expect(response.body.error).toBe(false);
+  });
+  test('Should return status 404 and no subscriptions message', async () => {
+    await Subscription.collection.deleteMany();
+    const response = await request(app).get(validRoute).send();
+    expect(response.status).toBe(404);
+    expect(response.body.data).toBeDefined();
```
Comment (LucianoAlarcon) [inline src/tests/subscriptions.test.js:40]: Please, change `toBeDefined()` to `toEqual([])`

Comment (jramire5) [reply src/tests/subscriptions.test.js:40]: In this one, or in every body.data?

File: src/tests/subscriptions.test.js [hunk 13]
```diff
@@ -1,5 +1,124 @@
...
+const validRoute = '/api/subscriptions';
+const mockSubscriptionId = subSeed[0]._id;
+const invalidId = '1234p';
+const notFoundId = '65694cb404b48dafdecfa44a';
+
+const mockSubscription = {
+  classId: '64623a627aebbd9653af45e1',
+  members: ['64623a627aebbd9653bf45e4', '64623a627bebbd9653af45e5'],
+  date: '2023-10-05T03:00:00.000+00:00',
+
```
[the remark this answers is not in the document]
Comment (jramire5) [reply src/tests/subscriptions.test.js:16]: Where is my friend Lint? Ok

**Review by LucianoAlarcon — APPROVED** [review]

**Comment by jramire5** [comment]
Thank you all!

**Pull request merged**